go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -140,7 +129,8 @@ func (npw *nodePortWatcher) AddService(service *kapi.Service) {
 }
 
 func (npw *nodePortWatcher) UpdateService(new, old *kapi.Service) {
-	if reflect.DeepEqual(new.Spec, old.Spec) {
+	// don't process services if there are no changes
+	if apiequality.Semantic.DeepEqual(new.Spec, old.Spec) && apiequality.Semantic.DeepEqual(new.Status, old.Status) {
```

Comment:
Is this fixing any actual observed bug? The commit comment mentions that "reflect deepequal doesn't consider defaults, apiequality does", but that shouldn't be relevant here, since `old` and `new` are both objects from the apiserver that have already had defaults filled in.

Reply:
not much into these AI machinery internals, and the comment is not helping, but it seems that for endpoints and slices, and in most parts of kubernetes is needed to avoid issues with empty fields https://github.com/kubernetes/kubernetes/pull/50583

Reply:
OK, so the problem is not with defaults, it's with fields being deserialized differently depending on how you receive the object.

Reply:
defaults vs empty thing too , right? I think that the API in kubernetes is moving towards using pointers everywhere to avoid that